extras/advanced-settings.md
- [Advanced Topics](#advanced-topics)
  - [Settings](#settings)
    - [Configuration files](#configuration-files)
    - [Basic structure](#basic-structure)
    - [Yes/No or “Similar values”](#yesno-or-similar-values)
    - [Ranges](#ranges)
    - [`observer`](#observer)
    - [`messaging`](#messaging)
    - [`cui`](#cui)
    - [`receivers`](#receivers)
    - [`dumphfdl`](#dumphfdl)
    - [`output`](#output)
    - [`db`](#db)
  - [Abbreviated Settings Examples](#abbreviated-settings-examples)
    - [tandem](#tandem)
    - [alpha web-888 roms](#alpha-web-888-roms)
    - [wide888](#wide888)
  - [Running as a Traffic Viewer (tbd)](#running-as-a-traffic-viewer-tbd)
  - [Running as a remote “node”](#running-as-a-remote-node)
  - [Experimental Conductor Options (Not Supported)](#experimental-conductor-options-not-supported)
    - [`type: static` (Experimental)](#type-static-experimental)
    - [`type: bfi` (Experimental)](#type-bfi-experimental)
  - [Experimental WebUI (Not Supported)](#experimental-webui-not-supported)
    - [Settings](#settings-1)
    - [URLs Served](#urls-served)
    - [Globe](#globe)
  - [ACARS Router Viewer (Experimental)](#acars-router-viewer-experimental)

# Advanced Topics

This document discusses some advanced configurations for HFDLObserver.

## Settings

The basic configuration generated by `hfdlobserver.sh configure` or at install time is good enough for the baseline use case: using a Web-888 device with no special hiccups. More advanced configurations are possible. This includes:

- different starting values for the CUI
- using different receivers instead of, or in addition to, a Web-888
- adding further outputs (to local AcarsRouter, AcarsHub, or other aggregators)

**Always keep a copy of your working settings.yaml before making alterations.**

### Configuration files

HFDLObserver uses only one configuration file. By default it is named `settings.yaml`. The app will look first in the current working directory, then the installation directory. You may also specify the settings file with the `--config` command line parameter.

The settings are stored in a [YAML]() formatted file. Much of this document is concerned with the structure of this file, and presumes at least a minimum understanding of the format. There are several good overviews for you to find. Only simple data structures are used.

As a reminder, key points for YAML:

- subsections are indented.
- indentation must be consistent
- indentation is done with spaces
- while strings/text may *usually* be left unquoted, it’s best to quote using either single or double quotes. Don’t quote numbers.
- a list of items begins with a `-` and an indent for properties on each item in the list.

### Basic structure

The settings.yaml file has several sections.

- `observer` - controls the central behaviour – managing receivers and gathering data
- `node` - controls behaviour when HFDLObserver is run in “node” mode to be managed by a remote “observer”. (More documentation on this mode is coming Real Soon Now.)
- `cui` - controls the behaviour of the Console UI, which is the “fancy” display

There are also sections that are used to build the configuration of the receivers that HFDLObserver controls.

- `receiver` - defines the top level configuration of the receiver
- `dumphfdl` - define the possible configurations of dumphfdl that are used by receivers
- `output` - defines possible outputs that can be used in receivers’ dumphfdl

### Yes/No or “Similar values”

Some properties take “Yes/No or Similar values”. means any of a number of different representations of yes or no.

- For yes: `"yes"`, `"true"`, `true`, `"y"`, or `1`
- For no: `"no"`, `"false"`, `false`, `"n"`, or `0`

### Ranges

Some properties accept a list of ranges.

``` yaml
property_name: [1000, 2000, [3000, 8000]]
```

A list of ranges is a list (`[]`) of either:
- a single explicit number; or
- a sublist expressing the closed interval of a range

In the example above, the exact values 1000, and 2000 are included, as is the range from 3000 to 8000 (inclusive).

### `observer`

#### `conductor`

This controls how frequencies are allocated to receivers.

``` yaml
  conductor:
    ranked_stations: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 13, 14, 15, 16]
    ignored_frequencies: [10000, [0, 3000]]
```

- `ranked_stations` - a list of ground station IDs from the systable.conf. It determines the relative importance of the active frequencies for each ground station from most to least important. HFDLObserver will cover active frequencies for stations earlier in the list before it allocates receivers to active frequencies for those later in it.
- `ignored_frequencies` - If you know that certain frequencies are unusable at your station, you can exclude them here. This is a list. Each entry in the list is either a single specific frequency, or is itself a list representing the closed interval `start` to `end`.
- `type` - You should never need to change this from the internally defaulted value. It is intended to aid in experimental extensions. Valid values: `diverse` (the default), `static`, and `bfi`.

#### `local_receivers`

This subsection describes the receivers that HFDLObserver *directly* controls – as opposed to “remote” nodes that manage additional receivers under HFDLObserver’s command. It is a list of objects. In theory, these could be quite complicated. In general, though, each entry will provide only a `name` and a `receiver` template to use for the receiver. The receiver object is described in detail in the `receivers` section later.

``` yaml
  local_receivers:
    - name: local-01
      receiver: web888
    - name: airspy-01
      receiver: airspyhf
```

Each local receiver needs an entry here. If you were to use a Web-888 device, you would need entries for all thirteen available web channels for that device. The configuration tool generates these for you.

(Any additional properties you add to a receiver object here will override any inherited from `receivers`, `dumphfdl`, or `outputs`)

#### `hfdl_listener`

This represents a lightweight endpoint for accepting HFDL packets in JSON format as emitted by dumphfdl. This is how HFDLObserver knows about packets that its receivers have received. If you are running in a stock configuration, there is little reason to include this section or alter the defaults.

``` yaml
  hfdl_listener:
    address: "127.0.0.1"
    port: 5540
    advertised_address: "192.168.0.42"
```

- `address` - The address to listen on. Specify `"0.0.0.0"` to listen to all available addresses your computer has.
- `port` - The UDP port to listen on. The port is advertised to any receivers that HFDLObserver is managing.
- `advertised_address` - When you use a generic address such as `"0.0.0.0"` as `address`, external/remote nodes will not know what address to contact. This is used for providing an explicit address to advertise to such nodes. If you aren’t running any “remote” nodes, there’s no reason to change this. If this is omitted and it is needed, HFDLObserver will try to guess, but it is very easy for it to be wrong.

#### `tracker`

This is another section you probably won’t need to change. It manages how ground stations frequencies are tracked.

``` yaml
  tracker:
    save_delay: 8
    state: "stations.state"
    station_files: ["systable.conf"]
    station_updates:
      - url: "https://api.airframes.io/hfdl/ground-stations"
        period: 131
      - url: "https://example.com/backup/list"
        period: 3600
```

- `save_delay` - A number in seconds. HFDLObserver will wait this many seconds after a change in a ground station’s frequencies before updating persistent data, or triggering a re-orchestration. This allows potentially several changes to be bundled together.
- `state` - A filename to cache the station active frequency data. This allows better recovery from network outages and restarts.
- `station_files` - This is a list of filenames/paths to the “systable.conf” files that dumphfdl uses. Usually there will be only one entry, and it defaults to the one provided in the repository. Unless you’re doing something weird, you won’t need to change this.

#### `station_updates`

This is a list of objects that tell HFDLObserver where to source “community” updates of active frequencies. The default uses [HFDLObserver](https://hfdl.observer/)’s own aggregator, but there are others. The format accepted for the data is a superset of the Airframes Ground Station API. You generally will not need to alter this.

Each station_update object contains two properties:
- `url` - the URL of the community data source
- `period` - how often to query it.

### `messaging`

Used to configure the control plane (messaging) for HFDLObserver. Unless you have port conflicts, you should never need to change this.

``` yaml
  messaging:
    host: "0.0.0.0"
    pub_port: 5559
    sub_port: 5560
```

- `host` - The address to use for ZMQ/TCP message brokering.
- `pub_port` - Used for nodes in HFDLObserver to publish ZMQ messages
- `sub_port` - Used for nodes in HFDLObserver to subscribe to ZMQ topics

### `cui`

This section configures the initial behaviour of the Console UI. It is currently fairly simple, but leaves the possibility open for further enhancement later.

#### `ticker`

This subsection configures the flowing “heat map” displayed at the top of the UI.

``` yaml
cui:
  ticker:
    bin_size: 60
    display_mode: "frequency"
    flexible_width: no
    show_all_active: yes
    show_active_line: yes
    show_confidence: yes
    show_targetting: no
```

- `bin_size` - packets are grouped into bins of this many seconds for display. Minimum value is `60`, maximum value is `3600`. The default is `60`
- `display_mode` - the initial kind of heat map displayed. The default is `frequency`. The valid values:
  - `frequency` - each row is a frequency, and the map shows packet counts for each bin for each frequency.
  - `band` - each row is a “MHz band”, which groups counts for all frequencies based on their integral MHz.
  - `station` - packets are grouped by the source or destination ground station, regardless of frequency.
  - `agent` - packets are grouped by the “agent” (dumphfdl’s “station”) representing the station name that would appear at airframes.io.
  - `receiver` - packets are grouped by the managed receiver (by name) that captured the packet. If packets come from a receiver that is not registered with HFDLObserver, they will show up as “unknown”.
- `flexible_width`: `yes`/`no` (or similar values, see above). When `yes`, the width of each bin column can vary depending on the packet count. This gives a precise number of packets, but can reduce the number of bins visible once you get to hundreds of packets per bin. This is most useful for users using large `bin_size`s. When `no`, each packet count is represented by a single character and the maximum number of bins is shown. The default is `no`.
- `show_all_active`: `yes`/`no` (or similar values, see above). When `yes`, all frequencies marked “active” even if there are no receivers assigned to the frequency or if no recent packets have been received on that frequency. The default is `no`.
- `show_active_line`: `yes`/`no` (or similar values, see above). When `yes`, the display will use a horizontal line to represent the period that an active frequency has been active (to the best of its knowledge). This helps illustrate when a ground station switches from one frequency to another. The default is `yes`.
- `show_confidence`: `yes`/`no` (or similar values, see above). When `yes`, the UI contains a symbol (in a circle) indicate how confident HFDLObserver is that this is an active frequency. There are some cases when squitters are not received directly from a ground station, and HFDLObserver relies on data from other ground stations. This can be somewhat out of date. The default is `yes`.
- `show_targetting`: `yes`/`no` (or similar values, see above). When `yes`, an additional symbol (in a square) indicates whether the given frequency is being actively targetted, is an “extra” that happens to be in the scanning range of a receiver, or is not tracked. The default is `no`. This is mostly used for debugging issues in the orchestrator.

### `receivers`

This section defines the types of receivers that can be added in `observer` § `local_receivers`. Any setting specified in the template can be overridden for a specific receiver, but its “path” has to match. There are examples of this in the **Abbreviated Settings Examples** section.

Each receiver template is defined in its own subsection, with a unique name. A receiver always includes a `decoder` section, that represents configuration for managing (and running) dumphfdl. Some *types* of Receiver also add a `client` section. This is used for `web-888` receivers, since they need a separate process to stream the IQ data from the device to the decoder/dumphfdl.

The basic structure of a receiver will look like this (lines with elipses represent settings that are discussed later, not literal lines):

``` yaml
receivers:
  web-888:
    type: "Web888ExecReceiver"
    channel_width: 12000
    client:
      type: "KiwiClientProcess"
      # ... client-specific properties ...
    decoder:
      type: "IQDecorderProcess"
      dumphfdl: "default"
      # ... decoder-specific properties ...
      output:
        - output: acarshub
          # ... output-specific properties ...
```

Some receiver *types* are predefined. If you have a subsection in your `receivers` section with the same name, it will inherit those defaults, so you only need to specify differences or additions. The defined receiver templates are:

- `web888` - the default receiver template for managing a connection and decoding of Web-888 channels.
- `airspyhf` - a receiver template for managing Airspy HF+ (and probably some other related) devices. The SoapyAirspyHF library must be installed for this to work. This is not done automatically during installation.
- `rx888mk2` - a receiver template for managing Web-888’s sibling device: RX-888 mk2. This depends on the Soapy driver contained in the ExtIO library, so that must be installed/built for this template to work.
- `rspdx+sdrplay` - a receiver template for managing SDRPlay RSPdx devices using the proprietary SDRPplay API. The proprietary drivers and the related SoapySDRPlay library must be installed/built for this template to work.

The following receiver templates are incomplete or experimental. Don’t expect them to work.

- `rsp1a+sdrplay` - an untested receiver template similar to `rspdx+sdrplay`, only for SDRPlay RSP1a devices.
- `rspdx+miri` - a broken receiver template for RSPdx using the open source Mirics Soapy driver. This driver does not actually support RSPdx, so this currently does not function.
- `pipe888` - a different way of managing Web-888 channels. (in this case, a shell manages the pipe rather than HFDLObserver directly)
- `dummy` - for development, this does nothing except receive an assignment, and pretend it’s doing something. For development.

#### `receiver` § `type: Web888ExecReceiver`

This is the basis for the default template `web888`. It uses a client process to fetch IQ data from the Web-888 device and a decoder process (dumphfdl) to process the signals and generate HFDL packets. It also allows you to specify the overall sample rate available.

``` yaml
  web888:
    type: "Web888ExecReceiver"
    channel_width: 12000
    client:
      type: "KiwiClientProcess"
      address: "web-888.local"
      port: 8073
      settle_time: 1
      quiet: false
      username: kiwi:hfdlobserver
      agc_files:
        "*": agc.yaml
        "21": agc_21M.yaml
    decoder:
      type: "IQDecoderProcess"
      dumphfdl: "default"
```

This is a fairly complete example representing the default definition of the `web888` template. In your settings file, you will only need to specify those settings you wish to override and the reasonable defaults will be applied to the others.

- `type` - this determines that this is the `Web888ExecReceiver` type.
- `channel_width` - each channel managed on receivers of this type will have this bandwidth (in samples/second or hz). The default (12000) is fine for most cases, but also see the **Abbreviated Settings Examples** “wide888” and “alpha web-888” for other uses.

##### `client`

- `type` - Technically, there could be different types of client, but only one (currently) makes sense for this Receiver type: `KiwiClientProcess`.
- `address` - the network name or address of the Web-888 device to use for this receiver. See the **Abbreviated Settings Example** “tandem” to see how this can be used to manage multiple Web-888 devices on the same network.
- `port` - the port on `address` where KiwiSDR web requests can be made. The default is almost always what you want.
- `settle_time` - a number in seconds representing the delay between stopping a previous use of the receiver, and restarting it (with new frequency assignments). This allows the device to reset the connection.
- `quiet` - setting this to `true` will reduce the log output of the client process a little.
- `username` - used by the Web-888 device to identify who’s using a particular channel. The default is one that works for the default installation. Password protected access is not supported.
- `agc_files` - this is a mapping of a frequency band to a **AGC YAML** that Web-888 can recognize. The keys for the mapping are the band’s frequency in MHz, and the value is the name of the file containing the AGC curve definition (see below). A key of `"*"` can also be specified; this serves as the default if a band is not provided or the file is not found. This is advanced tinkering, but can eke out slightly better reception if you can stand the near voodoo nature of these curves. (The useful bands are `"*"`, `2`, `3`, `4`, `5`, `6`, `8`, `10`, `11`, `13`, `15`, `17`, and `21`.)

##### `decoder`

- `type` - Technically, there could be different types of decoder, but only one (currently) makes sense for this Receiver type: `IQDecoderProcess`
- `dumphfdl` - Specifies the dumphfdl profile to use. See the **dumphfdl** section below.
- additional values may be specified here to override the values inherited from the dumphfdl profile.

#### `receiver` § `type: DirectReceiver`

A direct receiver does not require a process to retrieve IQ samples for dumphfdl. Dumphfdl access the device directly over USB or a similar connection. Generally, SoapySDR drivers are used, and can be configured by the `dumphfdl` profile chosen, or any overrides provided in the `receivers` template.

As dumphfdl access the radio directly, no `client` is required, and there is only a `decoder` section.

This example is the default definition of the `airspyhf` Receiver. It uses the `DirectReceiver` type:

``` yaml
  airspyhf:
    type: "DirectReceiver"
    decoder:
      type: "SoapySDRDecoder"
      settle_time: 1
      sample-rates: [912000, 768000, 650000, 456000, 384000, 228000, 192000]
      dumphfdl: "default"
      soapysdr:
        driver: "airspyhf"
        gain: None
```

Many of the entries have already been covered, but there are a couple of new ones.

- `sample-rates` - a list of the sample rates (samples per second, or Hz) supported by this profile. It needs to be values that the radio actually supports, but does not need to be all of them (for example, if you know your system has issues with higher sample rates, you could omit them). See “ranges” discussed above.
- `soapysdr` - this is a sub section that defines the SoapySDR arguments that `dumphfdl` will need to configure the radio. The properties in this section vary from driver to driver. You can check with `SoapySDRUtil --probe` to see what properties might be applicable here. The predefined HFDLObserver receivers have reasonable defaults for many of them.
- `soapysdr` § `driver` is common so that `dumphfdl` knows which SoapySDR driver/library to load for this radio.
- `soapysdr` § `gain` - for the `airspyhf` SoapySDR driver, the `gain` argument is accepted. If `None` (the default), AGC is enabled. Otherwise, it will be a value that is supported by the device (different devices will have different ranges).

#### `receiver` § `type: DummyReceiver`

A dummy development Receiver type for implementing the `dummy` template described above. Any settings are ignored. This is intended for development work and regular users can ignore it.

#### `receiver` § `type: Web888PipeReceiver`

The backing type for the `pipe888` template described above. It uses all the same settings as `Web888ExecReceiver`. It also may have incomplete functionality. It’s included here for some attempt at completeness.

### `dumphfdl`

This section defines the various profiles that can be used by receivers to configure the `dumphfdl` component. Similar to receivers, each profile is named. These names can be referenced in `reciever` § `decoder` sections, or other `dumphfdl` profiles (for a simplistic form of inheritance/extension).

``` yaml
dumphfdl:
  default:
    quiet: yes
    decoder_path: "dumphfdl"
    shoulder: 0.8
    system_table: "systable.conf"
    system_table_save: "systable_updated.conf"
    station_id: "EX-AMPL-E"
    airframes_enabled: yes
    statsd_server: "my-statsd-server:1234"
    output:
      - output: "acarshub"
      - output: "readsb"
        port: 9999
    packetlog: "/home/ducky/packet-logs/"
```

This is a fairly complete example of a basic `dumphfdl` profile named `default`. As noted above, special receiver types can add further properties to this. The additional configuration is usually handled as override properties in the `receiver` § `decoder` section.

- `quiet` - `yes`/`no` (or similar values, see above). If `yes`, only more important messages from `dumphfdl` will be relayed to `HFDLObserver`’s logs.
- `decoder_path` - if you have a custom build of `dumphfdl` and want to use it for in receivers that reference this profile, you can put the path here. The path is relative to the config file. It is recommended to use an absolute path if the executable is not on the PATH. The default just looks for `dumphfdl` on the system PATH.
- `shoulder` - this is the factor to apply to a receiver’s sample rate to determine the actual usable sample rate. Some radios introduce filters that attenuate some signals at the upper and lower ends of the sampling window. This parameter identifies how much of the sample window is considered “passed”. This is not needed for Web-888 profiles.
- `system_table` - passed to `dumphfdl` as the `--system-table` parameter.
- `system_table_save` - passed to `dumphfdl` as the `--system-table-save` parameter.
- `station_id` - passed to `dumphfdl` as the `--station-id` parameter. It is also used to identify packets collected by this “agent” within HFDLObserver. It is optional, but highly recommended. There is no default.
- `airframes_enabled` - `yes`/`no` (or similar values, see above). The default is `yes`. If this is `yes` and `station_id` is also defined, a `dumphfdl` output for airframes will automatically be generated. You can disable this if you are using a relay or router to pass packets on to airframes and don’t need HFDLObserver to communicate with airframes directly.
- `output` - a list of **output** profiles. These are described below. They operate in the same way as receivers and dumphfdl, in that you name an output (in this example `"acarshub"`), and can also override properties from the profile.

src/hfdl_observer/settings.py
# hfdl_observer/settings.py
# copyright 2025 Kuupa Ork <kuupaork+github@hfdl.observer>
# see LICENSE (or https://github.com/hfdl-observer/hfdlobserver/blob/main/LICENSE) for terms of use.
# TL;DR: BSD 3-clause
#

import itertools
import json
import os
import pathlib
from typing import Any, MutableMapping, Union

import yaml

import hfdl_observer.env as env
import hfdl_observer.util as util

base_path: pathlib.Path = pathlib.Path(os.getcwd())
registry: MutableMapping[str, dict] = {}


def dereference(ref_key: str, type_key: str, data: MutableMapping, *configs: MutableMapping) -> MutableMapping:
    try:
        key = data[ref_key]
    except KeyError:
        return data
    defaults = [config.get(type_key, {}).get(key, {}) for config in configs]
    return util.DeepChainMap({k: v for k, v in data.items() if k != ref_key}, *defaults)


def dereference_output(output: MutableMapping, *configs: MutableMapping) -> MutableMapping:
    try:
        key = output["output"]
    except KeyError:
        return output
    defaults = [config.get("outputs", {}).get(key, {}) for config in configs]
    return util.DeepChainMap({k: v for k, v in output.items() if k != "output"}, *defaults)


def dereference_dumphfdl(dumphfdl: MutableMapping, *configs: MutableMapping) -> MutableMapping:
    try:
        key = dumphfdl["dumphfdl"]
    except KeyError:
        return dumphfdl

    bases = [config.get("dumphfdl", {}).get(key, {}) for config in configs]
    complex_outputs: dict[str, list[MutableMapping]] = {}
    resolved_outputs = []
    source_outputs = [dumphfdl] + bases
    for output in itertools.chain.from_iterable(s["output"] for s in source_outputs if "output" in s):
        try:
            key = output["output"]
        except KeyError:
            resolved_outputs.append(output)
        else:
            complex_outputs.setdefault(key, []).append(output)
    for mappings in complex_outputs.values():
        chained = util.DeepChainMap(*mappings) if len(mappings) > 1 else mappings[0]
        resolved_outputs.append(dereference_output(chained, *configs))
    outputs = {"output": resolved_outputs}

    declared = {k: v for k, v in dumphfdl.items() if k != "output"}
    defaults = [{k: v for k, v in base.items() if k != "output"} for base in bases]
    return util.DeepChainMap(outputs, declared, *defaults)


def dereference_decoder(decoder: MutableMapping, *configs: MutableMapping) -> MutableMapping:
    if "dumphfdl" in decoder:
        return dereference_dumphfdl(decoder, *configs)
    return decoder


def dereference_receiver(receiver: MutableMapping, *configs: MutableMapping, level: int = 0) -> MutableMapping:
    # print(f'dereference receiver {receiver.get('name', 'anon')} = {receiver.get('receiver', 'n/a')}')
    try:
        key = receiver["receiver"]
    except KeyError:
        return receiver
    if level > 3:  # really naive way to break cycles. But really, you shouldn't be doing this much nesting.
        return receiver
    templates = [config.get("receivers", {}).get(key, {}) for config in configs]
    sources = [receiver] + templates
    _decoder = util.DeepChainMap(*[s["decoder"] for s in sources if "decoder" in s])
    decoder = {"decoder": dereference_decoder(_decoder, *configs)}

    declared = {k: v for k, v in receiver.items() if k != "decoder"}
    templated = [dereference_receiver(template, *configs, level=level + 1) for template in templates]

    return util.DeepChainMap(decoder, declared, *templated)


def dereference_receivers(receivers: list, *configs: MutableMapping) -> list:
    return [dereference_receiver(receiver, *configs) for receiver in receivers]


def chained(key: str, *configs: MutableMapping) -> MutableMapping:
    sources: list[MutableMapping] = [config[key] for config in configs if key in config]
    if len(sources) > 1:
        return util.DeepChainMap(*sources)
    elif sources:
        return sources[0]
    else:
        return {}


class DeprecatedSettingsError(Exception):
    pass


def load(filepath: Union[str, pathlib.Path]) -> MutableMapping:
    path = pathlib.Path(filepath)
    if not path.is_absolute():
        path = pathlib.Path(os.getcwd()) / path

    global base_path
    base_path = path.parent
    env.base_path = base_path

    global registry
    registry = yaml.safe_load(path.read_text())
    if "config" in registry:
        raise DeprecatedSettingsError()
    for parent_key in ["observer", "node"]:
        parent = registry.get(parent_key, {})
        if "local_receivers" in parent:
            parent["local_receivers"] = dereference_receivers(parent["local_receivers"], registry, defaults)
        else:
            parent["local_receivers"] = dereference_receivers(
                defaults[parent_key]["local_receivers"], registry, defaults
            )
    _globals = globals()
    for key in ["observer", "cui", "node", "viewer", "aggregator", "db"]:
        _globals[key] = chained(key, registry, defaults)
    return registry


defaults: dict[str, Any] = {
    "observer": {
        "conductor": {
            "type": "diverse",
            # "ranked_stations": [],
            # "ignored_frequencies": []
        },
        "hfdl_listener": {
            "address": "127.0.0.1",
            "port": 5540,
            # "advertised_address": ...,
        },
        "tracker": {
            "save_delay": 8,
            "state": "stations.state",
            "station_files": ["systable.conf"],
            "station_updates": [{"url": "https://hfdl.observer/active.json", "period": 61}],
        },
        "messaging": {
            "host": "0.0.0.0",
            "pub_port": 5559,
            "sub_port": 5560,
        },
        "local_receivers": [{"receiver": "web888", "name": f"web888-{i:02}"} for i in range(1, 14)],
    },
    "cui": {
        "ticker": {
            "bin_size": 60,
            "display_mode": "frequency",
            "show_all_active": False,
            "show_active_line": True,
            "show_confidence": True,
            "show_targetting": False,
            "flexible_width": False,
        },
        "secondary_displays": [],
    },
    "node": {"local_receivers": [], "messaging": {"host": "0.0.0.0", "pub_port": 5559, "sub_port": 5560}},
    "viewer": {"messaging": {"host": "0.0.0.0", "pub_port": 5559, "sub_port": 5560}},
    "dumphfdl": {
        "default": {
            "quiet": True,
            "decoder_path": "dumphfdl",
            "shoulder": 0.8,
            "system_table": "systable.conf",
            "system_table_save": "systable_updated.conf",
            "output": [],
        }
    },
    "outputs": {
        "hfdl_observer": {"format": "json", "protocol": "udp", "address": "hfdl.observer", "port": 5542},
        "acars_router": {"format": "json", "protocol": "udp", "address": "acars_router.local", "port": 5556},
        "acarshub": {"format": "json", "protocol": "udp", "address": "acarshub.local", "port": 5556},
        "readsb": {"format": "basestation", "protocol": "tcp", "address": "readsb.local", "port": 30009},
    },
    "receivers": {
        "web888": {
            "type": "Web888ExecReceiver",
            # "channel_width", 12000
            "client": {
                "address": "web-888.local",
                "port": 8073,
                "type": "KiwiClientProcess",
                "settle_time": 1,
                "quiet": False,
                "username": "kiwi:hfdlobserver",
                "agc_files": {
                    "*": "agc.yaml",
                    "2": "agc-02M.yaml",
                    "3": "agc-03M.yaml",
                    "4": "agc-04M.yaml",
                    "5": "agc-05M.yaml",
                    "6": "agc-06M.yaml",
                    "8": "agc-08M.yaml",
                    "10": "agc-10M.yaml",
                    "11": "agc-11M.yaml",
                    "13": "agc-13M.yaml",
                    "15": "agc-15M.yaml",
                    "17": "agc-17M.yaml",
                    "21": "agc-21M.yaml",
                },
            },
            "decoder": {"type": "IQDecoderProcess", "dumphfdl": "default"},
        },
        "pipe888": {"type": "Web888PipeReceiver", "client": {"type": "KiwiClient"}, "decoder": {"type": "IQDecoder"}},
        "dummy": {"type": "DummyReceiver"},
        "acars_router": {"type": "PullReceiver"},
        "airspyhf": {
            "type": "DirectReceiver",
            "decoder": {
                "type": "SoapySDRDecoder",
                "settle_time": 1,
                "sample-rates": [912000, 768000, 650000, 456000, 384000, 228000, 192000],
                "dumphfdl": "default",
                "soapysdr": {"driver": "airspyhf", "gain": None},
            },
        },
        "rx888mk2": {
            "type": "DirectReceiver",
            "decoder": {
                "type": "RX888mk2Decoder",
                "settle_time": 1,
                "sample-rates": [2000000, 4000000, 8000000],
                "dumphfdl": "default",
                "soapysdr": {"driver": "SDDC"},
                "gain": 89,
            },
        },
        "rspdx+sdrplay": {
            "type": "DirectReceiver",
            "decoder": {
                "type": "SoapySDRDecoder",
                "settle_time": 9,
                "sample-rates": [
                    62500,
                    96000,
                    125000,
                    192000,
                    250000,
                    384000,
                    500000,
                    768000,
                    1000000,
                    [2000000, 3000000],
                ],
                "dumphfdl": "default",
                "soapysdr": {"driver": "sdrplay"},
                "device-settings": {
                    "rfnotch_ctrl": False,
                    "dabnotch_ctrl": False,
                    "agc_setpoint": -14,
                    "biasT_ctrl": False,
                    "rfgain_sel": 0,
                },
            },
        },
        "rsp1a+sdrplay": {
            "type": "DirectReceiver",
            "decoder": {
                "type": "SoapySDRDecoder",
                "settle_time": 10,
                "sample-rates": [1300000, 1536000, 2048000, [2000000, 5500000]],
                "dumphfdl": "default",
                "soapysdr": {"driver": "sdrplay"},
                "device-settings": {
                    "rfnotch_ctrl": False,
                    "dabnotch_ctrl": False,
                    "agc_setpoint": -14,
                    "biasT_ctrl": False,
                    "rfgain_sel": 0,
                },
            },
        },
        "rsp1a+miri": {
            "type": "DirectReceiver",
            "decoder": {
                "type": "SoapySDRDecoder",
                "sample-rates": [1300000, 1536000, 2048000, 4000000, 5000000],
                "dumphfdl": "default",
                "soapysdr": {"driver": "soapyMiri", "bufflen": 65536, "buffers": 64, "asyncBuffs": 32},
                "device-settings": {"flavour": "SDRplay", "biastee": False},
            },
        },
    },
    "db": {
        "uri": "file:mem1?mode=memory&cache=shared",
        "horizon": 1,
    },
}
observer: MutableMapping = {}
cui: MutableMapping = {}
node: MutableMapping = {}
viewer: MutableMapping = {}
aggregator: MutableMapping = {}
db: MutableMapping = {}


def deepchainmap_representer(dumper: yaml.SafeDumper, dcm: util.DeepChainMap) -> yaml.nodes.MappingNode:
    return dumper.represent_dict(dcm.dict())


def get_dumper() -> type[yaml.SafeDumper]:
    safe_dumper = yaml.SafeDumper
    safe_dumper.add_representer(util.DeepChainMap, deepchainmap_representer)
    return safe_dumper


if __name__ == "__main__":
    import sys

    if len(sys.argv) > 1:
        filename = sys.argv[1]
    else:
        filename = "custom.yaml"
    loaded = load(filename)
    resolved = dict(loaded)
    for key in ["observer", "cui", "node", "viewer"]:
        resolved[key] = chained(key, loaded, defaults)
    print(yaml.dump(loaded, Dumper=get_dumper()))
    # out = yaml.safe_dump([d.dict() for d in observer['local_receivers']])
    # print(out)
    print(json.dumps(dict(resolved), indent=4, default=lambda o: o.dict() if hasattr(o, "dict") else list(o)))
